@xen-orchestra/backups-cli/index.js

```diff
@@ -0,0 +1,354 @@
+      // no longer needs to be checked
+      delete vhdParents[child]
+
+      // check the ancestors first
+      const grandparent = vhdParents[parent]
+      if (grandparent !== undefined) {
+        deleteIfOrphan(grandparent, parent)
+      }
+
+      if (!vhds.has(parent)) {
+        vhds.delete(child)
+
+        console.warn('Error while checking VHD', child)
+        console.warn('  missing parent', parent)
+        force && console.warn('  deleting…')
+        console.warn('')
+        force && deletions.push(handler.unlink(child))
+      }
+    }
+
+    // > A property that is deleted before it has been visited will not be
+    // > visited later.
+    // >
+    // > -- https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/for...in#Deleted_added_or_modified_properties
+    for (const child in vhdParents) {
+      deleteIfOrphan(vhdParents[child], child)
+    }
+
+    await Promise.all(deletions)
+  }
+
+  const [jsons, xvas] = await readDir(vmDir).then(entries => [
+    entries.filter(_ => _.endsWith('.json')),
+    new Set(entries.filter(_ => _.endsWith('.xva'))),
+  ])
+
+  await asyncMap(xvas, async path => {
+    // check is not good enough to delete the file, the best we can do is report
+    // it
+    if (!(await isValidTar(path))) {
+      console.warn('Potential broken XVA', path)
+      console.warn('')
+    }
+  })
+
+  const unusedVhds = new Set(vhds)
+  const unusedXvas = new Set(xvas)
+
+  await asyncMap(jsons, async json => {
+    const metadata = JSON.parse(await fs.readFile(json))
+    const { mode } = metadata
+    if (mode === 'full') {
+      const linkedXva = resolve(vmDir, metadata.xva)
+
+      if (xvas.has(linkedXva)) {
+        unusedXvas.delete(linkedXva)
+      } else {
+        console.warn('Error while checking backup', json)
+        console.warn('  missing file', linkedXva)
+        force && console.warn('  deleting…')
+        console.warn('')
+        force && (await handler.unlink(json))
+      }
+    } else if (mode === 'delta') {
+      let linkedVhds = metadata.vhds
+      linkedVhds = Object.keys(linkedVhds).map(key =>
+        resolve(vmDir, linkedVhds[key])
+      )
+
+      if (linkedVhds.every(_ => vhds.has(_))) {
+        linkedVhds.forEach(_ => unusedVhds.delete(_))
+      } else {
+        console.warn('Error while checking backup', json)
+        linkedVhds
+          .filter(_ => !vhds.has(_))
+          .forEach(linkedVhd => {
+            console.warn('  missing file', linkedVhd)
+          })
+        force && console.warn('  deleting…')
+        console.warn('')
+        force && (await handler.unlink(json))
+      }
+    }
+  })
+
+  // TODO: parallelize by vm/job/vdi
+  const unusedVhdsDeletion = []
+  {
+    // VHD chains (as list from child to ancestor) to merge indexed by last
+    // ancestor
+    const vhdChainsToMerge = { __proto__: null }
+
+    const toCheck = new Set(unusedVhds)
+
+    const getUsedChildChainOrDelete = vhd => {
```

Comment:
*Un*used?

Reply:
Nope, this returns the used child-chain which we have to merge, might be unclear though…